## Homework (variant 1): Comparison of pos-taggers

Natalia Semenova, MKL-171
In this task I'm trying to compare pos-taggers for german. I use as data TIGER corpus to evalute accuracy and simple text-file with some news article for evaluation of tagger work duration. As pos-tagger I've chosen:
1. ClassifierBasedGermanTagger (link https://github.com/ptnplanet/NLTK-Contributions/tree/master/ClassifierBasedGermanTagger)
2. spacy pos-tagger (de_core_news_sm-2.0.0 model link https://github.com/explosion/spacy-models/releases/tag/de_core_news_sm-2.0.0) 
3. StanfordPOSTagger (link https://nlp.stanford.edu/software/tagger.shtml)

### ClassifierBasedGermanTagger

In [76]:
import nltk
import re
from datetime import datetime
from nltk.corpus.reader.conll import *


In [77]:
#open a file with german text for processing
with open('deu-text.txt','r', encoding = "utf-8") as f:
    text = f.read()

In [78]:
print(text)

In den USA verdichten sich die Anzeichen für einen nahenden Rauswurf des nationalen Sicherheitsberaters General Herbert Raymond McMaster. Die "Washington Post" berichtet, die Ablösung sei ausgemachte Sache. US-Präsident Donald Trump wolle den Wechsel nur möglichst wenig erniedrigend gestalten, so das Blatt unter Berufung auf gleich fünf Quellen im Weißen Haus. Die "New York Times" hatte zuvor Ähnliches berichtet und auch den Namen von Stabschef und Ex-General John Kelly zu den Personalspekulationen hinzugefügt. McMaster, so mögliche Überlegungen, könnte ins Militär zurückkehren und zum Vier Sterne-General befördert werden. Seine Nachfolge im Weißen Haus ist offen. Als Kandidat wird auch der ehemalige Botschafter der USA bei den Vereinten Nationen gehandelt, der knallharte Falke John Bolton. Die Sprecherin des Weißen Hauses, Sarah Sanders, teilte auf Twitter mit, es gebe "keinen Wechsel" im Nationalen Sicherheitsrat. Sie habe mit Trump und McMaster gesprochen und diese hätten "entgegen

In [79]:
# maybe we will need tokenization for our pos-taggers, just let's have this function
def tokenize(one_new):
    tokens = one_new.replace(' - ', ' ').replace('. ', ' ')
    tokens = re.sub(r'\d+', '', tokens)
    umlya = re.compile(r'[\w]+')
    all_tokens = umlya.findall(tokens)
    return all_tokens


In [80]:
y = tokenize(text)
y

['In',
 'den',
 'USA',
 'verdichten',
 'sich',
 'die',
 'Anzeichen',
 'für',
 'einen',
 'nahenden',
 'Rauswurf',
 'des',
 'nationalen',
 'Sicherheitsberaters',
 'General',
 'Herbert',
 'Raymond',
 'McMaster',
 'Die',
 'Washington',
 'Post',
 'berichtet',
 'die',
 'Ablösung',
 'sei',
 'ausgemachte',
 'Sache',
 'US',
 'Präsident',
 'Donald',
 'Trump',
 'wolle',
 'den',
 'Wechsel',
 'nur',
 'möglichst',
 'wenig',
 'erniedrigend',
 'gestalten',
 'so',
 'das',
 'Blatt',
 'unter',
 'Berufung',
 'auf',
 'gleich',
 'fünf',
 'Quellen',
 'im',
 'Weißen',
 'Haus',
 'Die',
 'New',
 'York',
 'Times',
 'hatte',
 'zuvor',
 'Ähnliches',
 'berichtet',
 'und',
 'auch',
 'den',
 'Namen',
 'von',
 'Stabschef',
 'und',
 'Ex',
 'General',
 'John',
 'Kelly',
 'zu',
 'den',
 'Personalspekulationen',
 'hinzugefügt',
 'McMaster',
 'so',
 'mögliche',
 'Überlegungen',
 'könnte',
 'ins',
 'Militär',
 'zurückkehren',
 'und',
 'zum',
 'Vier',
 'Sterne',
 'General',
 'befördert',
 'werden',
 'Seine',
 'Nachfolge',
 '

In [81]:
# for reading our corpus 
corp = nltk.corpus.ConllCorpusReader('.', 'tiger_release_aug07.corrected.16012013.conll09',
                                     ['ignore', 'words', 'ignore', 'ignore', 'pos'],
                                     encoding='utf-8')

In [82]:
import random,math
#we need to separate data for training and testing
tagged_sents = corp.tagged_sents()
i = math.floor(len(tagged_sents)*0.2)
testing_sentences = tagged_sents[0:i]
training_sentences = tagged_sents[i:]

In [83]:
#now we train it (we can save model with the help of pickle module, but I didn't)
from ClassifierBasedGermanTagger import ClassifierBasedGermanTagger
tagger = ClassifierBasedGermanTagger(train=training_sentences) 

#### Evaluation (accuracy & time)

In [84]:
accuracy = tagger.evaluate(testing_sentences)

In [85]:
accuracy

0.9393353727898218

In [89]:
%timeit result_1 = tagger.tag(y)

508 ms ± 73.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [88]:
tagger.tag(y)

[('In', 'APPR'),
 ('den', 'ART'),
 ('USA', 'NE'),
 ('verdichten', 'VVFIN'),
 ('sich', 'PRF'),
 ('die', 'ART'),
 ('Anzeichen', 'NN'),
 ('für', 'APPR'),
 ('einen', 'ART'),
 ('nahenden', 'ADJA'),
 ('Rauswurf', 'NN'),
 ('des', 'ART'),
 ('nationalen', 'ADJA'),
 ('Sicherheitsberaters', 'NN'),
 ('General', 'NN'),
 ('Herbert', 'NE'),
 ('Raymond', 'NE'),
 ('McMaster', 'NE'),
 ('Die', 'ART'),
 ('Washington', 'NE'),
 ('Post', 'NE'),
 ('berichtet', 'VVFIN'),
 ('die', 'ART'),
 ('Ablösung', 'NN'),
 ('sei', 'VAFIN'),
 ('ausgemachte', 'ADV'),
 ('Sache', 'NN'),
 ('US', 'NE'),
 ('Präsident', 'NN'),
 ('Donald', 'NE'),
 ('Trump', 'NE'),
 ('wolle', 'VMFIN'),
 ('den', 'ART'),
 ('Wechsel', 'NN'),
 ('nur', 'ADV'),
 ('möglichst', 'ADV'),
 ('wenig', 'PIAT'),
 ('erniedrigend', 'ADJD'),
 ('gestalten', 'VVINF'),
 ('so', 'ADV'),
 ('das', 'ART'),
 ('Blatt', 'NN'),
 ('unter', 'APPR'),
 ('Berufung', 'NN'),
 ('auf', 'APPR'),
 ('gleich', 'ADV'),
 ('fünf', 'CARD'),
 ('Quellen', 'NN'),
 ('im', 'APPRART'),
 ('Weißen', 'ADJ

In [90]:
start_time = datetime.now()
tagger.tag(y)
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

Duration: 0:00:00.537752


### spacy
https://github.com/explosion/spacy-models/releases/tag/de_core_news_sm-2.0.0 - model for german

In [36]:
import spacy

In [37]:
nlp = spacy.load('de_core_news_sm-2.0.0/de_core_news_sm-2.0.0/de_core_news_sm/de_core_news_sm-2.0.0', disable=['parcer','ner'])

In [40]:
doc = nlp(text)

In [41]:
doc

In den USA verdichten sich die Anzeichen für einen nahenden Rauswurf des nationalen Sicherheitsberaters General Herbert Raymond McMaster. Die "Washington Post" berichtet, die Ablösung sei ausgemachte Sache. US-Präsident Donald Trump wolle den Wechsel nur möglichst wenig erniedrigend gestalten, so das Blatt unter Berufung auf gleich fünf Quellen im Weißen Haus. Die "New York Times" hatte zuvor Ähnliches berichtet und auch den Namen von Stabschef und Ex-General John Kelly zu den Personalspekulationen hinzugefügt. McMaster, so mögliche Überlegungen, könnte ins Militär zurückkehren und zum Vier Sterne-General befördert werden. Seine Nachfolge im Weißen Haus ist offen. Als Kandidat wird auch der ehemalige Botschafter der USA bei den Vereinten Nationen gehandelt, der knallharte Falke John Bolton. Die Sprecherin des Weißen Hauses, Sarah Sanders, teilte auf Twitter mit, es gebe "keinen Wechsel" im Nationalen Sicherheitsrat. Sie habe mit Trump und McMaster gesprochen und diese hätten "entgegen

In [43]:
def space_pos_tag(doc):
    for word in doc:
        print(word.text, word.tag_, spacy.explain(word.tag_))
    return doc

#### Evaluation (accuracy & time)

to evaluate accuracy I used de_pud-ud-test.conllu (link https://github.com/UniversalDependencies/UD_German-PUD) cause I didn't really figure out how to evaluate this model in other way (my fault). In cmd I used "python -m spacy convert" to convert conll into SpaCy's JSON format and than made  "python -m spacy evaluate" (link https://spacy.io/api/cli#evaluate), but result is pretty weird (something like POS 20.9). I add a picture result.png in Morphological disambiguation. I think I did smth wrong :/

In [44]:
from datetime import datetime
start_time = datetime.now()
m = space_pos_tag(doc)
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

In ADJD adjective, adverbial or predicative
den ART definite or indefinite article
USA NE proper noun
verdichten VVFIN finite verb, full
sich PRF reflexive personal pronoun
die ART definite or indefinite article
Anzeichen NN noun, singular or mass
für APPR preposition; circumposition left
einen ART definite or indefinite article
nahenden ADJA adjective, attributive
Rauswurf NN noun, singular or mass
des ART definite or indefinite article
nationalen ADJA adjective, attributive
Sicherheitsberaters NN noun, singular or mass
General NN noun, singular or mass
Herbert NE proper noun
Raymond NE proper noun
McMaster NE proper noun
. $. sentence-final punctuation mark
Die ART definite or indefinite article
" $. sentence-final punctuation mark
Washington NE proper noun
Post NE proper noun
" $. sentence-final punctuation mark
berichtet VVFIN finite verb, full
, $, comma
die ART definite or indefinite article
Ablösung NN noun, singular or mass
sei VAFIN finite verb, auxiliary
ausgemachte ADJA adj

### StanfordPOSTagger

In [45]:
from nltk.tag.stanford import StanfordPOSTagger

In [46]:
import os
java_path = "C:\\Program Files\\Java\\jdk-9.0.4\\bin\\java.exe"
os.environ['JAVAHOME'] = java_path

from nltk.tag.stanford import StanfordPOSTagger
path_to_model = "stanford-postagger-full-2018-02-27\stanford-postagger-full-2018-02-27\models\german-hgc.tagger"
path_to_jar = "stanford-postagger-full-2018-02-27\stanford-postagger-full-2018-02-27\stanford-postagger-3.9.1.jar"
tagger_stan=StanfordPOSTagger(path_to_model, path_to_jar)
tagger_stan.java_options='-mx4096m'          ### Setting higher memory limit for long sentences
sentence = 'Wir sing gleich alt, aber verschieden groß.'
print(tagger_stan.tag(sentence.split()))

[('Wir', 'PPER'), ('sing', 'VVFIN'), ('gleich', 'ADV'), ('alt,', 'ADV'), ('aber', 'ADV'), ('verschieden', 'ADJD'), ('groß.', 'ADJD')]


In [47]:
accuracy_3 = tagger_stan.evaluate(testing_sentences)

In [48]:
accuracy_3

0.9286613581896694

In [125]:
%timeit result_3 = tagger_stan.tag(y)
result_3

9.88 s ± 803 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


[('In', 'APPR'),
 ('den', 'ART'),
 ('USA', 'NE'),
 ('verdichten', 'VVFIN'),
 ('sich', 'PRF'),
 ('die', 'ART'),
 ('Anzeichen', 'NN'),
 ('für', 'APPR'),
 ('einen', 'ART'),
 ('nahenden', 'ADJA'),
 ('Rauswurf', 'NN'),
 ('des', 'ART'),
 ('nationalen', 'ADJA'),
 ('Sicherheitsberaters', 'NN'),
 ('General', 'NN'),
 ('Herbert', 'NE'),
 ('Raymond', 'NE'),
 ('McMaster', 'NE'),
 ('Die', 'ART'),
 ('Washington', 'NE'),
 ('Post', 'NN'),
 ('berichtet', 'VVFIN'),
 ('die', 'ART'),
 ('Ablösung', 'NN'),
 ('sei', 'VAFIN'),
 ('ausgemachte', 'ADJA'),
 ('Sache', 'NN'),
 ('US', 'NN'),
 ('Präsident', 'NN'),
 ('Donald', 'NE'),
 ('Trump', 'NE'),
 ('wolle', 'VMFIN'),
 ('den', 'ART'),
 ('Wechsel', 'NN'),
 ('nur', 'ADV'),
 ('möglichst', 'ADV'),
 ('wenig', 'ADV'),
 ('erniedrigend', 'ADJD'),
 ('gestalten', 'VVFIN'),
 ('so', 'ADV'),
 ('das', 'ART'),
 ('Blatt', 'NN'),
 ('unter', 'APPR'),
 ('Berufung', 'NN'),
 ('auf', 'APPR'),
 ('gleich', 'ADV'),
 ('fünf', 'CARD'),
 ('Quellen', 'NN'),
 ('im', 'APPRART'),
 ('Weißen', 'ADJ

In [119]:
start_time = datetime.now()
tagger_stan.tag(y)
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

Duration: 0:00:08.689433


So, we can see that the best accuracy was performed by the 1 tagger (ClassifierBasedGermanTagger, accuracy = 0.939335) and it works very fast (0:00:00.537752). But Stanford pos-tagger works also good (accuracy = 0.928661) and slower (0:00:08.689433).